drivers/ethernet/eth_w5500.c

```diff
@@ -0,0 +1,586 @@
+	offset %= mem_size;
+	addr = mem_start + offset;
+
+	if (offset + len > mem_size) {
+		remain = (offset + len) % mem_size;
+		len = mem_size - offset;
+	}
+
+	ret = w5500_spi_write(dev, addr, buf, len);
+	if (ret || !remain) {
+		return ret;
+	}
+
+	return w5500_spi_write(dev, mem_start, buf + len, remain);
+}
+
+static int w5500_command(struct device *dev, uint8_t cmd)
+{
+	uint8_t reg;
+	uint64_t end = z_timeout_end_calc(K_MSEC(100));
+
+	w5500_spi_write(dev, W5500_S0_CR, &cmd, 1);
+	do {
+		w5500_spi_read(dev, W5500_S0_CR, &reg, 1);
+		if (end - z_tick_get() <= 0) {
+			return -EIO;
+		}
+
+		k_msleep(1);
+	} while (reg != 0);
+
+	return 0;
+}
+
+static int w5500_tx(struct device *dev, struct net_pkt *pkt)
+{
+	struct w5500_runtime *ctx = dev->data;
+	uint16_t len = net_pkt_get_len(pkt);
+	uint16_t offset;
+	uint8_t off[2];
+	int ret;
+
+	w5500_spi_read(dev, W5500_S0_TX_WR, off, 2);
+	offset = sys_get_be16(off);
+
+	if (net_pkt_read(pkt, ctx->buf, len)) {
+		return -EIO;
+	}
+
+	ret = w5500_writebuf(dev, offset, ctx->buf, len);
+	if (ret < 0) {
+		return ret;
+	}
+
+	sys_put_be16(offset + len, off);
+	ret = w5500_spi_write(dev, W5500_S0_TX_WR, off, 2);
+	w5500_command(dev, S0_CR_SEND);
+
+	return 0;
+}
+
+static void w5500_isr(struct device *dev)
+{
+	uint8_t ir;
+	uint8_t mask = 0;
+	uint8_t header[2];
+	uint8_t tmp[2];
+	uint16_t rx_buf_len;
+	uint16_t rx_len;
+	uint16_t off;
+	struct w5500_runtime *ctx = dev->data;
+	const struct w5500_config *config = dev->config;
+	struct net_pkt *pkt;
+
+	while (true) {
+		k_sem_take(&ctx->int_sem, K_FOREVER);
+
+		w5500_spi_read(dev, W5500_S0_IR, &ir, 1);
+		if (!ir) {
+			goto done;
+		}
+
+		w5500_spi_write(dev, W5500_S0_IR, &ir, 1);
+
+		if (ir & S0_IR_SENDOK) {
+			LOG_DBG("TX Done");
+		}
+
+		if (ir & S0_IR_RECV) {
+			int read_len;
+			int reader;
+			struct net_buf *pkt_buf = NULL;
+
+			/* disable interrupt */
+			mask = 0;
+			w5500_spi_write(dev, W5500_SIMR, &mask, 1);
+			w5500_spi_read(dev, W5500_S0_RX_RSR, tmp, 2);
+			rx_buf_len = sys_get_be16(tmp);
+
+			if (rx_buf_len == 0) {
+				goto done;
+			}
+
+			w5500_spi_read(dev, W5500_S0_RX_RD, tmp, 2);
+			off = sys_get_be16(tmp);
+
+			w5500_readbuf(dev, off, header, 2);
+			rx_len = sys_get_be16(header) - 2;
+
+			pkt = net_pkt_rx_alloc_with_buffer(ctx->iface, rx_len,
+				AF_UNSPEC, 0, K_MSEC(config->timeout));
+			if (!pkt) {
+				eth_stats_update_errors_rx(ctx->iface);
+				goto done;
+			}
+
+			pkt_buf = pkt->buffer;
+
+			read_len = rx_len;
+			reader = off + 2;
+
+			do {
+				size_t frag_len;
+				uint8_t *data_ptr;
+				size_t frame_len;
+
+				data_ptr = pkt_buf->data;
+
+				frag_len = net_buf_tailroom(pkt_buf);
+
+				if (read_len > frag_len) {
+					frame_len = frag_len;
+				} else {
+					frame_len = read_len;
+				}
+
+				w5500_readbuf(dev, reader, data_ptr, frame_len);
+				net_buf_add(pkt_buf, frame_len);
+				reader += frame_len;
+
+				read_len -= frame_len;
+				pkt_buf = pkt_buf->frags;
+			} while (read_len > 0);
```

Comment:
hum, I don't like this way of writing packet into net_pkt but that reminds me we don't provide helpers to avoid a in-between buffer if net_pkt_write is being used.

That said, since you already have a ctx->buf ... you could just use it and avoid this loop.

Reply:
@tbursztyka @jukkar This wasn't about the `pkt_buf`. The `Sn_RX_RD` can be as large as `Sn_RXBUF_SIZE` (receive buffer size), which is configured as 16KB. So even `net_pkt_rx_alloc_with_buffer` can allocate as large as the size received, to read the data from `W5500_Sn_RX_MEM_START` we need to statically allocate (`ctx->buf`) as `Sn_RXBUF_SIZE` (16KB) , which I felt too large. That's the main reason for `net_buf` looping.

Reply:
Ok, anyway we will need to provide function on net_pkt api so this kind of loop could be avoided.

Reply:
I will open a enhancement request in issues for this. Thanks. Marking it as resolved here.